### DISCUSSION TITLE
Using StyleX in a prop based component library

### DISCUSSION THREAD

**Discussion by BMCwebdev** [facebook/stylex#341] [Q&A] [answered]
First, want to thank you for putting in all the time you have and spending time here in this discussion area chatting with us all as we learn this library. 

When I started building the component library I am now a maintainer of, I took a lot of inspiration from Mark Dalgleish. As such, a large mental model of the library is that users interact with it almost entirely via props. A typical example would be something like:
`<Button size=“large” color=“red” secondary icon=“house”>Go Home</Button>`
Most of the work of writing CSS has been abstracted away. In fact, almost all the components will not accept any user CSS.

It seems this system doesn’t entirely map itself to StyleX, but please let me know if you see it differently. 

As I understand it, I cannot pass in the user’s prop values to the StyleX CSS as that is considered dynamic. Though it is known at build time. For example, currently the component may be built and used like this. We would have some tokens defined
```tsx
// tokens.js
export const Tokens = {
  colors: {
    black: '#000000',
    blue: '#007cbc',
    red: '#DB1016',
  },
  sizes: {
    1x: '4px',
    2x: '8px',
  },
}
```
Then the component 
```tsx
// Button.jsx
import Tokens from ‘tokens’;

// these props would be TypeScript typeof keyof token.js. to constrain.
const Button = ({
  bgColor,
  padding,
  // add a default style to this prop
  textColor = 'black',
}) => {
  const buttonStyles = css`
    background-color: ${Tokens.colors[bgColor]};
    color: ${Tokens.colors[textColor]};
    padding: ${Tokens.sizes[padding]};
  `,
  return (
    <button css={buttonStyle}>{children}</button>
  )
}
```
And Dev Experience
```tsx
// App.jsx
import Button from ‘Button’;

const App = () => {
  return <Button bgColor=“red” padding=“2x”>Hello</Button>
}
```

And this would be the StyleX version, where the user must create their own style object.

Some tokens
```tsx
// tokens.js
export const colors = stylex.defineVars({
  black: '#000000',
  blue: '#007cbc',
  red: '#DB1016',
});

export const sizes = stylex.defineVars({
  1x: '4px',
  2x: '8px',
});
```

The component
```tsx
// Button.tsx
import {colors, sizes} from 'tokens.js';

// add a default style
// unlike previous example, we do not have props where we can
// define a default value. Even if we did, stylex.create must be outside
// of the component, so it would not have access to that prop value.
// instead we create a stylex object.
const styles = stylex.create({
  default: {
    color: colors.black,
  },
});

// quick example of types to constrain accepted styles.
type Props = {
  customizedStyles?: StyleXStyles<{
    // assume I could use some keyof typeof instead of longhand?
    backgroundColor: colors.black | colors.blue | colors.red,
    color: colors.black | colors.blue | colors.red,
    padding: sizes.1x | sizes.2x,
  }>,
};

const Button = ({customizedStyles}) => {
  return (
    <button {...stylex.props(styles.default, customizedStyles)}>
      {children}
    </button>
  );
};
```

And Dev Experience
```tsx
// App.tsx
import Button from ‘Button’;

// instead of prop based, user would create and pass in a style object.
const devChosenStyles = stylex.create({
  backgroundColor: colors.blue,
  color: colors.black,
  padding: sizes.1x,
});

const App = () => {
  return <Button customizedStyles={devChosenStyles}>Hello</Button>
};
```

Do I have this right so far? That even though at build time we know what `color: ${Tokens.colors[color]};`  is, because the user passed that in via the prop, it’s still considered dynamic and therefore cannot use that pattern in StyleX?

Do you know of a way StyleX could still be used in a prop-based component library? It would be such a major shift from the way our component library is established it does make it hard to adopt, get buy in, etc.

That, right now, is my biggest question.

If you’re still reading this, thanks! I’m also curious about defining variables and then redefining them via themes.

It looks like when you make tokens with `defineVars` you cannot nest those definitions. The pattern appears to be creating a const each for `colors`, `spaces`, etc. Similarly when you create a theme with `createTheme` you can only redefine one variable object at a time.

What I am missing here is the ability to create a contract that encompasses all of the design systems tokens. And then being able to redefine that whole collection for a theme. As I understand it how it works now would lead to having a collection of themes to redefine all the tokens like `InternalColors`, `InternalSpaces` etc. instead of just one theme `Internal`

I _love_ this: “All variables in a variable group must be overridden when creating a theme. This choice has been to help catch accidental omissions. Even at the cost of occasional tedium.” 

Curious, how deep does this requirement go? Say a defineVars has something with a dark option
`primaryText: {default: 'black', [DARK]: 'white'},`
If the theme redefines this, but omits a `[DARK]` will there be a warning, or is that considered fine, and the important thing is that all keys are present and have a value?

One more question? Pseudo elements. We currently use Mark and Michael Taranto’s Capsize on all text nodes. That uses ::before and ::after quite extensively to do it’s magic. StyleX documentation recommends against the use of pseudo elements. I plan on testing this one out soon, but while I’m here just thought I would ask what the drawback is and why pseudo elements are discouraged.

Wow, you’re still here? Thanks again for reading all of this. Hope you may have some ideas. Excited to see how people start using StyleX. Congratulations again!

**Comment by nmn** [comment]
> Do I have this right so far? That even though at build time we know what color: ${Tokens.colors[color]}; is, because the user passed that in via the prop, it’s still considered dynamic and therefore cannot use that pattern in StyleX?

Yes.

> because the user passed that in via the prop

A prop is inherently dynamic since it can be different values in different scenarios. 

> Do you know of a way StyleX could still be used in a prop-based component library?

Yes you can you the variant pattern mentioned here: https://stylexjs.com/docs/learn/styling-ui/using-styles/#style-variants

In your example, it would look like this:

```tsx
// Button.tsx
import {colors, sizes} from 'tokens.js';

// add a default style
// unlike previous example, we do not have props where we can
// define a default value. Even if we did, stylex.create must be outside
// of the component, so it would not have access to that prop value.
// instead we create a stylex object.
const styles = stylex.create({
  default: { color: colors.black},
});
const colorVariants = stylex.create({
  black: {color: colors.black},
  blue: {color: colors.blue},
  red: {color: colors.red},
});
const bgVariants = stylex.create({
  black: {backgroundColor: colors.black},
  blue: {backgroundColor: colors.blue},
  red: {backgroundColor: colors.red},
});
const sizeVariants = stylex.create({
  '1x': {padding: sizes['1x']},
  '2x': {padding: sizes['2x']},
});

// quick example of types to constrain accepted styles.
type Props = {
  // I might be doing the Types wrong...
  bgColor: keyof typeof bgVariants,
  padding: keyof typeof sizeVariants,
  textColor?: keyof typeof colorVariants,
};

const Button = ({bgColor, padding, textColor}: Props) => {
  return (
    <button {...stylex.props(
      styles.default,
      bgVariants[bgColor],
      colorVariants[textColor],
      sizeVariants[padding],
    )}>
      {children}
    </button>
  );
};
```

With this approach, you may not need to define `tokens.js` at all depending how the values are used.

**Comment by BMCwebdev** [comment]
Hey, thanks so much for the feedback, incredibly helpful!

Cool, I can see how this would allow a prop-based component library use StyleX, and while you are right, this approach might negate the need to define tokens.js at all, that's kinda a drawback, yeah?

It would be nice to have all the tokens defined for the design system in one location (token.js), and it would be nice if that were the source of truth you can derive off of.

For example, in this approach you coded above, if we were to add a new color to the list of tokens.js, we would have to go into every component and add it as an option to each variant.

In the old approach, the tokens.js file dictated everything. The component derived the typeof keyof for the props off that one centralized location (providing the options to the user).

This approach looks to introduce another layer of maintenance and things could get out of sync.

I'll have to think about this a bit more. It's related to what I was mentioning about themes as well. It would be wonderful to have one `defineVars` that held all the design tokens, a contract for the design system. And then singular themes would have to follow that contract and supply their own values.

Thanks again Naman, I've been following you and this project for a long time and super happy for you.

**Reply by nmn** [reply, marked as answer]
> It would be nice to have all the tokens defined for the design system in one location (token.js), and it would be nice if that were the source of truth you can derive off of.

The styles defined with `styles`, `colorVariants`, `bgVariants` and `sizeVariants` can be defined and exported from a file called `tokens.js` and imported and used in various components.

Alternatively, the component can use variables for everything, and the variants can be defined as themes in a central location.

----

The tradeoff between the two approaches is that exporting styles directly is simpler and generates less CSS, but you have to control each component individually.

With themes, components can be themed for an entire UI sub-tree.

------

With StyleX almost every problem can be solved with architecture changes.
